### PR TITLE
Allows import to succeed when realm  names include whitespaces

### LINKED ISSUE #11921: Validation for realm name not containing space OLD DESCRIPTION: Realm import fails with "Base URL is not a valid URL; A redirect URI is not a valid URI"
We should ideally have validation for realm name not containing space as allowing space in the realm name causes unneccesary headaches. We should take migration into consideration as it is possible that people still use space in their old realm names (So likely import of the JSON file with `keycloakVersion` < version-where-we-fix-this-bug should be allowed with some warning in the server.log). See coment for the details: https://github.com/keycloak/keycloak/issues/11921#issuecomment-1127641525

### Describe the bug (Original description)

I created a new realm through the admin UI with name "Test Realm", and keep everything at default values, 
I then exported that realm with `./bin/kc.sh export --dir ./export `, and copied the "Test Realm-realm.json" into ./data/import/realm.json.
I then deleted "Test Realm" through the admin UI.

If I run `./bin/kc.sh start --import-realm`, the import fails with:
```
Appending additional Java properties to JAVA_OPTS: -Djgroups.dns.query=localhost
2022-05-10 08:45:55,784 INFO  [org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: FrontEnd: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin: <request>, Port: -1, Proxied: false
2022-05-10 08:45:57,384 WARN  [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000569: Unable to persist Infinispan internal caches as no global state enabled
2022-05-10 08:45:57,401 WARN  [org.infinispan.PERSISTENCE] (keycloak-cache-init) ISPN000554: jboss-marshalling is deprecated and planned for removal
2022-05-10 08:45:57,471 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
2022-05-10 08:45:57,820 INFO  [org.infinispan.CONTAINER] (keycloak-cache-init) ISPN000128: Infinispan version: Infinispan 'Triskaidekaphobia' 13.0.8.Final
2022-05-10 08:45:57,998 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000078: Starting JGroups channel `ISPN`
2022-05-10 08:46:00,218 INFO  [org.jgroups.protocols.pbcast.GMS] (keycloak-cache-init) a-2chnjtenrvouy-4750: no members discovered after 2002 ms: creating cluster as coordinator
2022-05-10 08:46:00,228 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000094: Received new cluster view for channel ISPN: [a-2chnjtenrvouy-4750|0] (1) [a-2chnjtenrvouy-4750]
2022-05-10 08:46:00,233 INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000079: Channel `ISPN` local address is `a-2chnjtenrvouy-4750`, physical addresses are `[10.247.56.85:7800]`
2022-05-10 08:46:00,803 INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: a-2chnjtenrvouy-4750, Site name: null
2022-05-10 08:46:04,194 INFO  [org.infinispan.CLUSTER] (main) ISPN000080: Disconnecting JGroups channel `ISPN`
2022-05-10 08:46:04,313 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2022-05-10 08:46:04,314 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to import realm: Test Realm
2022-05-10 08:46:04,314 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Invalid client account: Base URL is not a valid URL; A redirect URI is not a valid URI
2022-05-10 08:46:04,315 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
```

Keycloak was built with:
```
ENV KC_DB=mysql
ENV KC_HTTP_RELATIVE_PATH=/auth
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true
```

It is started with the following environment variables:
```
JAVA_OPTS_APPEND="-Djgroups.dns.query=localhost"
KC_HOSTNAME_STRICT=false
KC_HOSTNAME_STRICT_HTTPS=false
KC_HTTP_PORT=8080
KC_HTTP_ENABLED=true
KC_DB_URL_DATABASE=keycloak_test
KC_DB_URL_HOST=xxx
KC_DB_USERNAME=xxx
KC_DB_USERNAME=xxx
```

### Version

18.0

### Expected behavior

It should be possible to import realm that was exported with keycloak.

### Actual behavior

Import fails

### How to Reproduce?

1. Create new "Test Realm" with default settings
2. Export the realm with `./bin/kc.sh export --dir ./export`
3. Copy exported "Test Realm-realm.json" to ./data/import/realm.json
4. Delete "Test Realm" through the admin UI
5. Start keycloak with "./bin/kc.sh start --import-realm"

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by DGuhr**
cc @pedroigor @vmuzikar @abstractj As said, this does the trick, but I am not entirely sure about this solution (alternative could be to e.g. encode the URLs before giving them to the validation), so draft state for now. Test will follow, too. Could you please give me your thoughts on this one? Fixes it for both, wildfly and quarkus.

Closes #11921

### CODE REVIEW

File: services/src/main/java/org/keycloak/validation/DefaultClientValidationProvider.java [hunk 1]
```diff
@@ -184,7 +184,7 @@ private void checkUri(FieldMessages field, String url, ValidationContext<ClientM
         }
 
         try {
-            URI uri = new URI(url);
+            URI uri = new URI(url.replace(" ","%20"));
```
Comment (abstractj) [inline services/src/main/java/org/keycloak/validation/DefaultClientValidationProvider.java:187]: @DGuhr I just wonder if it wouldn't make more sense to use:

```
url.replaceAll("\\s+","%20")
```

**Comment by DGuhr** [comment]
@vmuzikar from the answer in the issue it seems to be a regression from Keycloak 6, yes. Last PR that touched these was titled `KEYCLOAK-14306 OIDC redirect_uri allows dangerous schemes resulting in potential XSS` and seems to come from you, the title is part of why I am not sure this is the right solution ;)

File: services/src/main/java/org/keycloak/validation/DefaultClientValidationProvider.java (hunk 1 above)
Comment (DGuhr) [reply services/src/main/java/org/keycloak/validation/DefaultClientValidationProvider.java:187]: definitely makes sense _when we go this way_, but lets answer the bigger question first before moving on i'd say :)

Comment (DGuhr) [reply services/src/main/java/org/keycloak/validation/DefaultClientValidationProvider.java:187]: (see also the answer to @vmuzikar i gave below)

**Comment by vmuzikar** [comment]
@DGuhr Thank you for the clarification!

IMHO loosening the validation in URLs in general is not the way to go. We could maybe instead try to sanitize the URLs on Realm Import, before any validation happens.

**Comment by DGuhr** [comment]
@vmuzikar updated the PR now to sanitize it at the right point, at least for the --import-realm at startup for quarkus. have to find the spots for wildfly, and I think `sanitizeClientUrlsForImport` will go to the RealmRepresentation itself as a method. But this solution now worked perfectly for realms with one or multiple whitespaces inside. more tomorrow.

**Pull request closed without merging**